### PR TITLE
setup containers with releases

### PR DESCRIPTION
**Pull request by InoMurko**
This is part 1 of work that needs to be done to move our deployment towards OTP releases.
All images are based on Alpine 3.8 - https://github.com/omisego-images/docker-elixir-omg
Integration tests run towards custom alpine with geth and solc. It would be good to move away from this model though.

Processes and flows have been adjusted in a way similar to eWallet... thanks @sirn !!! But with separate repositories so that we're not coupled with their work.

### CODE REVIEW

File: apps/omg_child_chain/config/dev.exs [hunk 1]
```diff
@@ -3,5 +3,4 @@ use Mix.Config
 
 config :omg_child_chain,
   ethereum_events_check_interval_ms: 500,
-  block_queue_eth_height_check_interval_ms: 1_000,
-  coordinator_eth_height_check_interval_ms: 1_000
```
Comment (InoMurko) [inline apps/omg_child_chain/config/dev.exs]: it was not used at all

File: Dockerfile.child_chain [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        bash \
+        libressl \
+        libressl-dev \
+        lksctp-tools
+
+COPY rootfs /
+
+# USER directive is not being used here since privileges are dropped via
+# s6-setuigid in /entrypoint. s6-overlay is required to be run as root.
+ARG user=child_chain
```
Comment (jbunce) [inline Dockerfile.child_chain]: Are we rolling with `child_chain`? I know it's not related to this PR but there's a lot of inconsistency in this project between Childchain/child_chain/Child Chain etc

File: Dockerfile.child_chain [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+ARG release_version
+
+ADD _build/prod/rel/child_chain/releases/${release_version}/child_chain.tar.gz /app
+RUN chown -R "${uid}:${gid}" /app
+WORKDIR /app
+
+# Child Chain app is using PORT environment variable to determine which port to run
+# the application server.
+ENV PORT 9656
```
Comment (jbunce) [inline Dockerfile.child_chain]: This isn't going to be something that changes often, and the reference to the variable is used directly below. I'd remove this.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+#
+
+docker-child_chain-prod:
+	docker run --rm -it \
+		-v $(PWD):/app \
+		-v $(IMAGE_BUILD_DIR)/deps:/app/deps \
+		-u root \
+		--entrypoint /bin/sh \
+		$(IMAGE_BUILDER) \
+		-c "cd /app && make build-child_chain-prod"
```
Comment (jbunce) [inline Makefile]: Not sure on this mixture of hyphens and underscores. Makes me think childchain is better.

File: Dockerfile.child_chain (hunk 2 above)
Comment (InoMurko) [reply Dockerfile.child_chain]: I don't know what to do! :D

File: Makefile (hunk 4 above)
Comment (InoMurko) [reply Makefile]: I don't know what to do at this stage. A was told by @pdobacz the correct naming is two words.

**Comment by jbunce** [comment]
Will this still allow for ENVs to be collected at runtime? Presumably this clears the way for Distillery releases too?

File: Dockerfile.child_chain (hunk 3 above)
Comment (InoMurko) [reply Dockerfile.child_chain]: https://github.com/omisego/elixir-omg/blob/master/apps/omg_rpc/config/prod.exs#L7

**Comment by InoMurko** [comment]
> Will this still allow for ENVs to be collected at runtime? Presumably this clears the way for Distillery releases too?

Yes @jbunce the work that applies ENVs has been done via Config Providers (more about them here https://hexdocs.pm/distillery/extensibility/config_providers.html). In code, they can be observed as `OMG.Eth.ReleaseTasks.SetContract`. These modules (`https://github.com/omisego/elixir-omg/blob/master/rel/config.exs#L42`) kick in just before the application release starts, does what you tell it to do and applies the configuration.

File: Dockerfile.child_chain (hunk 2 above)
Comment (pnowosie) [reply Dockerfile.child_chain]: There was a discussion and resulted to `child chain` is a correct form. So in this context with underscore.

File: Makefile (hunk 4 above)
Comment (pdobacz) [reply Makefile]: `child-chain` is two words, so...

File: Dockerfile.child_chain (hunk 2 above)
Comment (pdobacz) [reply Dockerfile.child_chain]: > lot of inconsistency in this project between Childchain/child_chain/Child Chain

I'd say we're pretty consistent with ChildChain/child_chain/Child chain. Two words all of these. I just searched: `Childchain` is very rare, `Rootchain` doesn't show up, there's several `childchains` and `rootchains`. We can groom this at some point.

Comment (InoMurko) [reply Dockerfile.child_chain]: I agree. I don't think I've made a naming mistake in this PR. Everything is `child_chain`!

File: .circleci/config.yml [hunk 5]
```diff
@@ -281,49 +228,45 @@ jobs:
...
-      - image: omisegoimages/elixir-omg-circleci:v1.8-20190129-02
-        environment:
-          MIX_ENV: dev
-    working_directory: ~/repo
+  publish_child_chain:
+    executor: metal_child_chain
+    steps:
+      - checkout
+      - run: make docker-child_chain CHILD_CHAIN_IMAGE_NAME=$CHILD_CHAIN_IMAGE_NAME
+      #- run: sh .circleci/ci_publish.sh
```
Comment (pdobacz) [inline .circleci/config.yml]: is there a plan for the commented lines here or just some leftovers to be deleted? If the former, maybe worth to comment

File: apps/omg_child_chain/config/dev.exs (hunk 1 above)
Comment (pdobacz) [reply apps/omg_child_chain/config/dev.exs]: :+1: whoops! missed that or maybe a git-merge fail? thx for catching!

File: mix.exs [hunk 6]
```diff
@@ -10,9 +10,7 @@ defmodule OMG.Umbrella.MixProject do
       deps: deps(),
       preferred_cli_env: [
         coveralls: :test,
-        "coveralls.detail": :test,
-        "coveralls.post": :test,
-        "coveralls.html": :test,
+        "coveralls.circle": :test,
```
Comment (pdobacz) [inline mix.exs]: I use `coveralls.details` and `.html` sometimes locally, please leave those in. 

`.post` is pre-circle era so can go

File: .circleci/config.yml (hunk 5 above)
Comment (InoMurko) [reply .circleci/config.yml]: yep, that comes in as part 2 when I setup deployment!

**Pull request merged**